File: fixtures/project-state/tools.go
// SPDX-License-Identifier: GPL-3.0-only
// Run from repos/sdk: go run ../../fixtures/project-state/tools.go > ../../fixtures/project-state/tools.json
package main

import (
	"context"
	"encoding/json"
	"os"
	"strings"
	"time"

	sdk "github.com/gratefulagents/sdk/pkg/agentsdk"
	ps "github.com/gratefulagents/sdk/pkg/agentsdk/projectstate"
	tools "github.com/gratefulagents/sdk/pkg/agentsdk/tools/projectstate"
)

type step struct {
	Name    string `json:"name"`
	Input   any    `json:"input"`
	Format  string `json:"format"`
	Save    string `json:"save,omitempty"`
	Output  any    `json:"output"`
	IsError bool   `json:"is_error"`
}

func main() {
	dir, err := os.MkdirTemp("", "state-tool-fixture")
	if err != nil {
		panic(err)
	}
	defer os.RemoveAll(dir)
	store, err := ps.NewFilesystemStore(ps.FilesystemOptions{StateDir: dir, ProjectID: "tool-contract"})
	if err != nil {
		panic(err)
	}
	catalog := tools.Tools(store, "agent")
	definitions := []any{}
	byName := map[string]sdk.Tool{}
	for _, tool := range catalog {
		byName[tool.Name()] = tool
		definitions = append(definitions, map[string]any{"name": tool.Name(), "description": tool.Description(), "input_schema": json.RawMessage(tool.InputSchema()), "read_only": tool.IsReadOnly(), "requires_approval": tool.NeedsApproval()})
	}
	steps := []step{
		{Name: "task_create", Input: map[string]any{"title": "blocker"}, Save: "$a"},
		{Name: "task_create", Input: map[string]any{"title": "dependent", "priority": 0}, Save: "$b"},
		{Name: "task_link", Input: map[string]any{"id": "$b", "depends_on": "$a"}},
		{Name: "task_ready", Input: map[string]any{}},
		{Name: "task_claim", Input: map[string]any{"id": "$a"}},
		{Name: "task_comment", Input: map[string]any{"id": "$a", "body": "note"}},
		{Name: "task_update", Input: map[string]any{"id": "$a", "labels": []string{"label"}}},
		{Name: "task_update", Input: map[string]any{"id": "$a", "labels": []string{}}},
		{Name: "task_close", Input: map[string]any{"id": "$a", "reason": "done"}},
		{Name: "task_ready", Input: map[string]any{}},
		{Name: "task_link", Input: map[string]any{"id": "$b", "depends_on": "$a", "action": "remove"}},
		{Name: "task_show", Input: map[string]any{"id": "$a"}},
		{Name: "memory_remember", Input: map[string]any{"content": "durable engineering preference", "kind": "semantic", "scope": "user", "tags": []string{"style"}}, Save: "$m"},
		{Name: "memory_update", Input: map[string]any{"id": "$m", "content": "compact engineering preference", "tags": []string{}}},
		{Name: "memory_list", Input: map[string]any{"kinds": []string{"semantic"}}},
		{Name: "memory_recall", Input: map[string]any{"query": "engineering"}},
		{Name: "memory_stats", Input: map[string]any{}},
		{Name: "prime_context", Input: map[string]any{"active_task_id": "$b", "memory_limit": nil, "ready_limit": nil}},
		{Name: "memory_delete", Input: map[string]any{"id": "$m"}},
		{Name: "memory_stats", Input: map[string]any{}},
	}
	steps = append(steps,
		step{Name: "task_create", Input: map[string]any{"title": "nullable", "description": nil, "priority": nil, "type": nil, "assignee": nil, "labels": nil, "depends_on": nil}, Save: "$c"},
		step{Name: "task_ready", Input: map[string]any{"limit": nil, "labels": nil, "assignee": nil, "include_assigned": nil}},
		step{Name: "task_update", Input: map[string]any{"id": "$c", "description": nil, "priority": nil, "title": nil, "labels": nil}},
		step{Name: "memory_remember", Input: map[string]any{"content": "nullable memory", "kind": nil, "scope": nil, "tags": nil, "task_ids": nil, "file_paths": nil, "source_run": nil}, Save: "$n"},
		step{Name: "memory_update", Input: map[string]any{"id": "$n", "content": nil, "kind": nil, "scope": nil, "tags": nil}},
		step{Name: "memory_list", Input: map[string]any{"limit": nil, "kinds": nil, "tags": nil}},
		step{Name: "memory_recall", Input: map[string]any{"query": "nullable", "limit": nil, "kinds": nil, "tags": nil}},
		step{Name: "memory_update", Input: map[string]any{"id": "$n", "tags": []any{nil, "keep"}}},
		step{Name: "task_update", Input: map[string]any{"id": "$c", "labels": []any{nil, "keep"}}},
		step{Name: "memory_stats", Input: nil},
		step{Name: "prime_context", Input: nil},
		step{Name: "task_create", Input: nil},
		step{Name: "task_create", Input: map[string]any{"title": ""}},
		step{Name: "memory_remember", Input: map[string]any{"content": ""}},
		step{Name: "memory_update", Input: map[string]any{"id": "missing", "content": "absent"}},
		step{Name: "memory_update", Input: map[string]any{"id": nil}},
		step{Name: "task_create", Input: map[string]any{"title": 42}},
		step{Name: "task_ready", Input: map[string]any{"limit": "oops"}},
	)
	for _, tool := range catalog {
		for _, input := range []any{[]any{}, "oops", 42, false} {
			steps = append(steps, step{Name: tool.Name(), Input: input})
		}
	}
	ids := map[string]string{}
	reverse := map[string]string{}
	for i := range steps {
		st := &steps[i]
		input := st.Input
		if fields, ok := st.Input.(map[string]any); ok {
			resolved := map[string]any{}
			for key, value := range fields {
				if str, ok := value.(string); ok && ids[str] != "" {
					value = ids[str]
				}
				resolved[key] = value
			}
			input = resolved
		}
		raw, _ := json.Marshal(input)
		result, err := byName[st.Name].Execute(context.Background(), raw, "call")
		if err != nil {
			panic(err)
		}
		st.IsError = result.IsError
		var output any
		st.Format = "json"
		if result.IsError || st.Name == "prime_context" {
			st.Format = "text"
			output = result.Content
		} else if err := json.Unmarshal([]byte(result.Content), &output); err != nil {
			panic(err)
		}
		if st.Save != "" {
			id := output.(map[string]any)["id"].(string)
			ids[st.Save] = id
			reverse[id] = st.Save
		}
		st.Output = normalize(output, reverse)
	}
	encoder := json.NewEncoder(os.Stdout)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(map[string]any{"definitions": definitions, "steps": steps}); err != nil {
		panic(err)
	}
}
func normalize(value any, ids map[string]string) any {
	switch v := value.(type) {
	case map[string]any:
		for key, item := range v {
			v[key] = normalize(item, ids)
		}
	case []any:
		for i, item := range v {
			v[i] = normalize(item, ids)
		}
	case string:
		if id := ids[v]; id != "" {
			return id
		}
		if _, err := time.Parse(time.RFC3339Nano, v); err == nil {
			return "<time>"
		}
		if strings.HasPrefix(v, "comment_") {
			return "<comment>"
		}
		for id, alias := range ids {
			v = strings.ReplaceAll(v, id, alias)
		}
		return v
	}
	return value
}

File: crates/adk-project-state/tests/tools.rs
use adk_core::*;
use adk_project_state::*;
use serde_json::{Value, json};
use std::sync::Arc;

struct NeverCancel;
impl Cancellation for NeverCancel {
    fn is_cancelled(&self) -> bool {
        false
    }
    fn cancelled(&self) -> BoxFuture<'_, ()> {
        Box::pin(std::future::pending())
    }
}
async fn invoke(tools: &[Arc<dyn Tool>], name: &str, arguments: Value, error: bool) -> String {
    let context = ToolContext {
        operation: Context {
            run_id: "tools-test".into(),
            cancellation: Arc::new(NeverCancel),
            deadline: None,
        },
        work_dir: Default::default(),
        policy: Default::default(),
        idempotency_key: None,
    };
    let output = tools
        .iter()
        .find(|t| t.definition().name == name)
        .unwrap()
        .execute(
            &context,
            ToolCall {
                id: "call".into(),
                name: name.into(),
                arguments,
            },
        )
        .await
        .unwrap();
    assert_eq!(output.is_error, error, "{name}: {output:?}");
    assert!(!output.should_pause);
    match &output.content[..] {
        [Content::Text { text }] => text.clone(),
        _ => panic!("non-text result"),
    }
}
async fn json_call(tools: &[Arc<dyn Tool>], name: &str, args: Value) -> Value {
    serde_json::from_str(&invoke(tools, name, args, false).await).unwrap()
}

#[tokio::test]
async fn baseline_state_tool_contracts_on_filesystem_and_sqlite() {
    for sqlite in [false, true] {
        let dir = tempfile::tempdir().unwrap();
        let open = || {
            let store = StoreOptions {
                project_id: "tool-contract".into(),
                ..Default::default()
            };
            Arc::new(if sqlite {
                ProjectStore::sqlite(SQLiteOptions {
                    path: dir.path().join("state.db"),
                    store,
                    ..Default::default()
                })
                .unwrap()
            } else {
                ProjectStore::filesystem(FilesystemOptions {
                    state_dir: dir.path().join("state"),
                    store,
                    ..Default::default()
                })
                .unwrap()
            })
        };
        let tools = tools::tools(open(), " agent ");
        assert_eq!(tools.len(), 15);
        for tool in &tools {
            let def = tool.definition();
            assert!(!def.requires_approval);
            assert_eq!(
                def.read_only,
                matches!(
                    def.name.as_str(),
                    "task_ready"
                        | "task_show"
                        | "memory_recall"
                        | "memory_list"
                        | "memory_stats"
                        | "prime_context"
                )
            );
        }
        let a = json_call(&tools, "task_create", json!({"title":"blocker"})).await;
        assert_eq!(a["priority"], 2);
        let b = json_call(
            &tools,
            "task_create",
            json!({"title":"dependent","priority":0}),
        )
        .await;
        assert_eq!(b["priority"], 0);
        json_call(
            &tools,
            "task_link",
            json!({"id":b["id"],"depends_on":a["id"]}),
        )
        .await;
        assert_eq!(
            json_call(&tools, "task_ready", json!({}))
                .await
                .as_array()
                .unwrap()
                .len(),
            1
        );
        let claimed = json_call(&tools, "task_claim", json!({"id":a["id"]})).await;
        assert_eq!(claimed["assignee"], "agent");
        let comment = json_call(&tools, "task_comment", json!({"id":a["id"],"body":"note"})).await;
        assert_eq!(comment["actor"], "agent");
        json_call(
            &tools,
            "task_update",
            json!({"id":a["id"],"labels":["label"]}),
        )
        .await;
        let cleared = json_call(&tools, "task_update", json!({"id":a["id"],"labels":[]})).await;
        assert!(cleared.get("labels").is_none());
        json_call(&tools, "task_close", json!({"id":a["id"],"reason":"done"})).await;
        assert_eq!(
            json_call(&tools, "task_ready", json!({})).await[0]["id"],
            b["id"]
        );
        json_call(
            &tools,
            "task_link",
            json!({"id":b["id"],"depends_on":a["id"],"action":"remove"}),
        )
        .await;
        let shown = json_call(&tools, "task_show", json!({"id":a["id"]})).await;
        assert_eq!(shown["status"], "closed");
        let mem = json_call(&tools,"memory_remember",json!({"content":"durable engineering preference","kind":"semantic","scope":"user","tags":["style"],"task_ids":[b["id"]]})).await;
        let updated = json_call(
            &tools,
            "memory_update",
            json!({"id":mem["id"],"content":"compact engineering preference","tags":[]}),
        )
        .await;
        assert_eq!(updated["scope"], "user");
        assert_eq!(updated["task_ids"], mem["task_ids"]);
        assert!(updated.get("tags").is_none());
        assert_eq!(
            json_call(&tools, "memory_list", json!({"kinds":["semantic"]}))
                .await
                .as_array()
                .unwrap()
                .len(),
            1
        );
        assert_eq!(
            json_call(&tools, "memory_recall", json!({"query":"engineering"})).await[0]["id"],
            mem["id"]
        );
        assert_eq!(
            json_call(&tools, "memory_stats", json!({})).await,
            json!({"total":1,"by_kind":{"semantic":1},"by_scope":{"user":1},"by_tag":{}})
        );
        assert!(
            invoke(
                &tools,
                "prime_context",
                json!({"active_task_id":b["id"]}),
                false
            )
            .await
            .contains("dependent")
        );
        drop(tools);
        let tools = tools::tools(open(), "agent");
        assert_eq!(
            json_call(&tools, "memory_list", json!({})).await[0]["content"],
            updated["content"]
        );
        json_call(&tools, "memory_delete", json!({"id":mem["id"]})).await;
        assert_eq!(
            json_call(&tools, "memory_stats", json!({})).await["total"],
            0
        );
        invoke(
            &tools,
            "memory_update",
            json!({"id":mem["id"],"content":"must not resurrect"}),
            true,
        )
        .await;
        invoke(&tools, "task_show", json!({"id":3}), true).await;
        invoke(&tools, "task_create", json!({"title":""}), true).await;
    }
}

fn normalize(value: &mut Value, ids: &std::collections::BTreeMap<String, String>) {
    match value {
        Value::Object(map) => {
            for value in map.values_mut() {
                normalize(value, ids);
            }
        }
        Value::Array(array) => {
            for value in array {
                normalize(value, ids);
            }
        }
        Value::String(text) => {
            if let Some(id) = ids.get(text) {
                *text = id.clone();
            } else if chrono::DateTime::parse_from_rfc3339(text).is_ok() {
                *text = "<time>".into();
            } else if text.starts_with("comment_") {
                *text = "<comment>".into();
            } else {
                for (id, alias) in ids {
                    *text = text.replace(id, alias);
                }
            }
        }
        _ => {}
    }
}

#[tokio::test]
async fn actual_go_tool_outputs_match_both_stores() {
    let fixture: Value =
        serde_json::from_str(include_str!("../../../fixtures/project-state/tools.json")).unwrap();
    for sqlite in [false, true] {
        let dir = tempfile::tempdir().unwrap();
        let opts = StoreOptions {
            project_id: "tool-contract".into(),
            ..Default::default()
        };
        let store = if sqlite {
            ProjectStore::sqlite(SQLiteOptions {
                path: dir.path().join("state.db"),
                store: opts,
                ..Default::default()
            })
            .unwrap()
        } else {
            ProjectStore::filesystem(FilesystemOptions {
                state_dir: dir.path().join("state"),
                store: opts,
                ..Default::default()
            })
            .unwrap()
        };
        let tools = tools::tools(Arc::new(store), "agent");
        assert_eq!(
            serde_json::to_value(
                tools
                    .iter()
                    .map(|tool| tool.definition())
                    .collect::<Vec<_>>()
            )
            .unwrap(),
            fixture["definitions"]
        );
        let covered: std::collections::BTreeSet<_> = fixture["steps"]
            .as_array()
            .unwrap()
            .iter()
            .map(|step| step["name"].as_str().unwrap())
            .collect();
        assert_eq!(covered.len(), 15);
        assert!(
            fixture["steps"]
                .as_array()
                .unwrap()
                .iter()
                .any(|step| step["is_error"] == true)
        );
        let mut ids = std::collections::BTreeMap::new();
        let mut reverse = std::collections::BTreeMap::new();
        for step in fixture["steps"].as_array().unwrap() {
            let mut input = step["input"].clone();
            normalize(&mut input, &ids);
            let is_error = step["is_error"].as_bool().unwrap();
            let output = invoke(&tools, step["name"].as_str().unwrap(), input, is_error).await;
            let mut result = if step["format"] == "text" {
                Value::String(output)
            } else {
                serde_json::from_str(&output).unwrap()
            };
            if is_error
                && step["output"]
                    .as_str()
                    .is_some_and(|s| s.starts_with("Invalid input:"))
            {
                assert!(
                    result.as_str().unwrap().starts_with("Invalid input:"),
                    "{result}"
                );
                continue;
            }
            if let Some(save) = step["save"].as_str() {
                let id = result["id"].as_str().unwrap().to_owned();
                ids.insert(save.to_owned(), id.clone());
                reverse.insert(id, save.to_owned());
            }
            normalize(&mut result, &reverse);
            assert_eq!(result, step["output"], "{} sqlite={sqlite}", step["name"]);
        }
    }
}
